### ISSUE TITLE
Swicth between sinks on run

### ISSUE THREAD

**Issue by cangermi** [gabime/spdlog#1162]
Hi, 

I have a server and 'n' number of different clients. I would like to log every single client's log file into a different folder. In order to do that I need 'n' number of logger. Instead of creating 'n' number of logger I created 'n' number of sinks. Everytime a request comes I would like to switch sinks or I just want to enable 1 sink for each request not all of them. My solution was creating another logger with all my sinks inside of it and everytime a request comes I change the my main loggers sink. But I do not think this method is thread safe. My server can get average 200 requests at the same time. Each one of this requests can belong to the same client or different clients. Actually I can put a scope lock inside of my logging method but if I get 200 requests from different clients every log has to wait the one that comes before it. If you have a thread safe solution  or a different solution to solve this problem I would like hear. Thanks.

**Comment by tt4g** [comment]
For your information.

It may not be a solution, but I have used the logger alone for each task referring to the wiki: https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-manually

The task is designed to provide a dedicated session context for each request, and by registering a logger in the context, dedicated log output was possible.

The following is a simplified code sample:

```cpp
#include <spdlog/spdlog.h>
#include <spdlog/sinks/stdout_sinks.h>

#include <memory>
#include <string>
#include <utility>
#include <vector>

class Session
{
public:

    explicit Session(std::size_t id, std::shared_ptr<spdlog::logger> logger)
            : id(id),
              logger(std::move(logger))
    {

    }

    std::size_t id;

    std::shared_ptr<spdlog::logger> logger;

};

// acquire task.
bool acquire_task(std::size_t& task_id);

void do_task(std::unique_ptr<Session> session)
{
    session->logger->info("start task.");

    // Do something ...
}

int main(int, char**)
{
    //// It is a sink shared by all tasks.
    std::vector<spdlog::sink_ptr> sinks({
            std::make_shared<spdlog::sinks::stdout_sink_mt>()
    });

    auto root_logger =
            std::make_shared<spdlog::logger>(
                    "root", sinks.begin(), sinks.end());
    spdlog::register_logger(root_logger);

    std::size_t task_id;
    while (acquire_task(task_id)) {
        //// I used the task id for the logger's name to identify the contents of the log.
        auto logger =
            std::make_shared<spdlog::logger>(
                    std::to_string(task_id) , sinks.begin(), sinks.end());
        //// Setup task's logger.
        logger->set_level(spdlog::level::info);
        logger->flush_on(spdlog::level::info);

        //// Don't call spdlog::register_logger(). When the task is over, discard the logger.
        // spdlog::register_logger(logger);

        auto session = std::make_unique<Session>(task_id, std::move(logger));

        do_task(std::move(session));
    }

    spdlog::shutdown();

    return 0;
}
```

In my case I share a sink. You may be able to achieve the purpose by preparing sinks for each one.

Note that this method doesn't work with async_logger. Because async_logger that is not managed by spdlog will be difficult to share thread pool.

**Comment by gabime** [comment]
A similar solution to @tt4g ‘s  solution is to create separate logger per client with single sink. 

This way you can get the right logger using spdlog::get() which is thread safe and reasonably fast.
Each time a new client is created, you create a new logger, register it with some id, and then use spdlog::get(id). Dont forget to un_register the logger once you’ve done with the client.

An even better approach is to store the logger along with your client object without registering it.But in any case separate logger per client is better than having multiple sinks in same logger.

**Comment by cangermi** [comment]
Than you for your responses.

I have an another question. There are continious requests coming to server from clients and every single one of them has to be logged. In that case, if I create my logger seperately and  each time a request comes I register and unresgister my loggers would not be cause performance issiues.Far as I know registering and unregistering operations are going to create and destroy logger threads. This is going to expensive for the machine resources.

**Comment by tt4g** [comment]
If you copy the generated logger (`std::shared_ptr<spdlog::logger>`) to the place you want to use in the request, you do not have to register.

Instead, just call `spdlog::logger::flush()` before destroying the logger.

**Comment by gabime** [comment]
> In that case, if I create my logger seperately and each time a request comes I register and unresgister my loggers would not be cause performance..

Sounds problematic for high rates. each time a logger is created/destroyed, a new file would be created/destroyed.. why not logger per client? not per request?

**Comment by cangermi** [comment]
> Sounds problematic for high rates. each time a logger is created/destroyed, a new file would be created/destroyed.. why not logger per client instead of per request?

I did exactly this. I have 'n' number of logers now. Before this I created 'n' number of sinks and everytime a request comes I change my main loggers sink from a another loger which I created just to hold my sinks. But if I get 200 request at the same time from different clients my logs will vanish. I just wanted to be sure of creating sinks per clients wont be a thread safe method.
